pkg/api/bzzchunk_test.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api_test
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/jsonhttp/jsonhttptest"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// TestChunkUpload uploads a chunk to an API that verifies the chunk according
+// to a given validator, then tries to download the uploaded data.
+// It does not test for the chunk validation logic but
+// rather that the chunk which is Put to the storage is passed through a validator
+// and that respective possible flows are handled correctly.
+func TestChunkUploadDownload(t *testing.T) {
+	resource := func(addr swarm.Address) string {
+		return "/bzz-chunk/" + addr.String()
+	}
+
+	validHash, err := swarm.ParseHexAddress("aabbcc")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	invalidHash, err := swarm.ParseHexAddress("bbccdd")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	validContent := []byte("bbaatt")
+	invalidContent := []byte("bbaattss")
+
+	validatorF := func(addr swarm.Address, data []byte) bool {
+		if !addr.Equal(validHash) {
+			return false
+
+		}
+		if !bytes.Equal(data, validContent) {
+			return false
+		}
+		return true
+	}
+
+	mockValidatingStorer := mock.NewValidatingStorer(validatorF)
+	client, cleanup := newTestServer(t, testServerOptions{
+		Storer: mockValidatingStorer,
+	})
+
+	defer cleanup()
```

Comment:
Extremely pedantic comment: this defer should be moved one line up.

Reply:
:D